man/lgp.1
.TH lgp 1 "2008-09-01" "logilab-packaging"

.SH NAME
lgp \- logilab packaging tool to ease projects' maintenance


.SH SYNOPSIS
.B lgp
.IR command
.IR [options]

.BR

.SH DESCRIPTION
lgp is a generic command to suit the logilab conventions about package management
Try lgp -h for more information.

.SH GENERAL OPTIONS

.B --verbose, -v
run silently without confirmation

.B --pkg_dir, -p
set a specific project directory

.B --revision, -r
set a specific revision or tag to build the debian package


.SH COMMANDS

.SS setup
prepare the chrooted distributions

.B --distrib, -d
the distribution targetted (e.g. stable, unstable)

.B --command, -c
command to run with pbuilder

.SS check
check and report problems about the code source directories and files

.B --distrib, -d
the distribution targetted (e.g. stable, unstable). Use 'all' for all known distributions

.B --include
force the inclusion of other check functions

.B --exclude
force the exclusion of other check functions

.B --list
return a list of all available check functions

.B --only
run only one single test

.SS prepare
run assurance quality tools on the source directory

.SS build
run the creation of a distribution package (source and/or binary)

.B --dist
where to put compilation results

.B --distrib, -d
the distribution targetted (e.g. stable, unstable). Use 'all' for all known distributions

.B --arch, -a
build for the requested debian architectures only

.B --orig-tarball
path to orig.tar.gz file

.B --keep-tmpdir
keep the temporary build directory

.B --post-treatments
compile packages with post treatments

.SS tag
put a release tag on source repository

.SS info
extract info from package configuration files

.SH EXAMPLE
.B lgp build -d etch -v <source_dir>

will run in verbose mode and build the etch packages for <source_dir>

.SH LICENCE
(c) 2006-2008 Logilab. General Public Licence.

.SH BUGS

.SH AUTHOR
Logilab <contact@logilab.fr>


ChangeLog
ChangeLog for logilab.devtools
==============================

2009-08-26 -- 0.16
    * lgp: can now build any Debian project (debian/)

    * lgp: new piuparts command (resurrected)

    * lgp: the --orig-tarball can manage URI directly (by urllib)

    * lgp: rudimentary multiprocess support for parallel build by architecture

    * lgp: don't stop prematurely if one of the package build fails

    * lgp: set a default distribution value by reading debian/changelog


2009-06-03 -- 0.15
    * lgp: add project command to print useful information

    * lgp: new suffix option for build command (incremental build)

    * #9031: lgp: include hooks to replace post-treatment

    * #5713: lgp: multi arch build (support i386 mode if linux32 is installed)

    * lgp: generate the build logfile in same directory as packages

    * lgp: simulate a local debian repository in creating a Packages.gz file in build result directory

    * lgp: add tag command to add list of tags with substitution

    * #9147: devtools : added lsprofcalltree module and command line script


2009-04-23  --  0.14
    * lgp: add login command to enter into chroot images

    * lgp: python-vcslib is no more a strict dependancy

    * lgp: update lgp pbuilderrc config file

    * lgp: add an intermediate facility step for debugging

    * vcslib: use new mercurial API (> 1.0)



2009-01-16  --  0.13
    * lgp: closes #4667: generate debian source package

    * lgp: #7994: run lgp check as pre-treatment for a build

    * vcslib: stay compatible with etch release (mercurial 0.9.1)



2008-10-14  --  0.12
    * lgp: can now manage pbuilder image to create package

    * changelog: support for debian NMU version

    * .pycoveragerc at the root of your project can specify omit and directory options



2008-06-11  --  0.11.1
    * various bugfix

    * vcslib: urlview support for HG

    * --new changelog: support for debian NMU version



2008-05-19  --  0.11.0
    * force manifest file regeneration

    * updated for mercurial API changes

    * fixed vcsagent.log_info method api: from_date and to_date should be
      local time tuple, return an iterator on CheckInInfo instances

    * fixed hg local time conversion

    * fixed cvs log output parsing

    * coverage disable settrace while running with its own trace
      function. This break doctests but avoid coverage problems...



2006-12-06  --  0.10.2
    * fixed silly NameError

    * moved optparse commands extension into logilab-common

    * changelog: fix to consider version.txt

    * updated for logilab-common 0.21

    * new hgstatus script



2006-08-04  --  0.10.1
    * fixed silly NameError



2006-08-03  --  0.10.0
    * fixed spelling of dependant/dependent

    * Move to new Python packaging policy

    * logilab.el: insert current year in copyright (#10779)

    * checkpackage: correctly ignore mercurial files (#14462)

    * cvslog: added man page (#8730)

    * preparedistrib: added retry option when step failed (#8862)

    * debianize: removed (#14473)

    * debian package refactor (#14476)

    * lgp: one program to rule them all (#14467)



2006-05-31  --  0.9.0
    * buildpackage: offers to test the packages with piuparts

    * preparedistrib: offers to copy the COPYING file from a set of
      known licenses 

    * buildpackage: fixed missing DEBUILDER environment variable bug

    * dos2unix: removal (unused, buggy, conflicts with debian package tofrodos)

    * debianize: fix the way to handle subpackage's __init__.py file (there
      is now a subpackage_master boolean property in the __pkginfo__.py which
      tells if a package is handling the __init__.py file, so now only one 
      subpackage should set this to True and the others should depends on 
      this package)

    * debianize: fix Uploader to Uploaders in control

    * pkginfo: fix debian handler detection

    * makedistrib: don't ask to tag package if an error occurs

    * vcslib: added mercurial support



2006-01-10  --  0.8.0
    * debianize:

    * updated to handle site-python installation with architecture
        independent package (implements #10119)

    * consider the optional debian_uploader variable in package
        information 

    * fix .examples files generation with debianize

    * fix templates: control.standard-version is now 3.6.2, updated FSF
        address in debian licenses files

    * fix package type detection in preparedistrib (fix #9917)



2005-07-29  --  0.7.0
    * fix dh -i/-a option handing in rules file generated by debianize
      (#9700)

    * no more empty lines annoying emacs in rules files generated by
      debianize  (#9710)

    * better use of .docs files / dh_installdocs and .examples files
      / dh_installexamples in debianize

    * better pbuilder support in buildeb (#1045)

    * tagpackage try to tag only distributed files (#674)

    * catch EOFError and KeyboardInterrupt when waiting for an input to
      avoid traceback

    * fix svn_agent.log_info message format



2005-07-19  --  0.6.7
    * fix call to pycoverage with missing arguments (close #9584)

    * makedistrib force manifest regeneration when building debian
      packages (close #9134)



2005-06-20  --  0.6.6
    * fix for apycot checkers (python2.2 compat + handle apycot >= 0.7)



2005-06-14  --  0.6.5
    * new options on checkpackage to select tests to apply

    * handle optional dependency on logilab-common for the -test package

    * fix svn tagging (both svn and cvs tags now include the
      project's name)

    * fix SGML catalog checking for registered DTDs with another language
      than EN



2005-02-25  --  0.6.4
    * fixed various bugs in lib.changelog

    * added write-permission check before checkin



2005-02-16  --  0.6.3
    * fix pycoverage to not include python file which are not in a
      subpackage (i.e. directory without __init__.py  file)

    * debianize: fix html doc detection, fix bug with py2.3 and optik


narval/lgp.py

import sys
import os, os.path as osp
import logging

from narvalbot.prototype import input, output
from narvalbot.elements import FilePath

import cubes.apycot
import apycotlib as apycot
from apycotlib.checkers import BaseChecker, AbstractFilteredFileChecker

STEP_DEBIANPKG = 3

from logilab.devtools.lgp import *

@apycot.apycotaction('lgp.check', 'CHECKEDOUT in elmt.done_steps')
def act_lgp_check(inputs):
    test = inputs['apycot']
    cwd = os.getcwd()
    os.chdir(test.project_path())
    try:
        checker, status = test.run_checker('lgp.check', inputs.get('options'))
    finally:
        os.chdir(cwd)
    return {}

@output('changes-files', 'isinstance(elmt, FilePath)', 'elmt.type == "debian.changes"', list=True)
@apycot.apycotaction('lgp.build', 'CHECKEDOUT in elmt.done_steps')
def act_lgp_build(inputs):
    test = inputs['apycot']
    checker, status = test.run_checker('lgp.build', inputs.get('options'))
    return {'changes-files': checker.debian_changes}

class LgpLogHandler(logging.Handler):
    def __init__(self, writer):
        logging.Handler.__init__(self)
        self.writer = writer
        self.path = None
        self.status = apycot.SUCCESS

    def emit(self, record):
        if record.levelno >= logging.CRITICAL:
            f = self.writer.fatal
            self.status = apycot.FAILURE
        elif record.levelno >= logging.ERROR:
            f = self.writer.error
            self.status = apycot.FAILURE
        elif record.levelno >= logging.WARNING:
            f = self.writer.warning
        elif record.levelno >= logging.INFO:
            f = self.writer.info
        else:
            f = self.writer.debug
        f(record.getMessage(), path=self.path)

class LgpCheckChecker(BaseChecker):
    """run tests on a project with lgp"""

    id = 'lgp.check'
    command = 'check'
    options_def = {
    }

    def do_check(self, test):
        cmd = LGP.get_command(self.command)
        handler = LgpLogHandler(self.writer)
        cmd.logger.addHandler(handler)
        exit_status = cmd.main_run(['-vv'], LGP.rcfile)
        if exit_status:
            self.writer.fatal('lgp %s exited with status %s', self.command, exit_status)
            self.set_status(apycot.ERROR)
        return apycot.SUCCESS

apycot.register('checker', LgpCheckChecker)

class LgpBuildChecker(BaseChecker):
    """build debian packages with lgp"""

    id = 'lgp.build'
    command = 'build'
    options_def = {
        'lgp_build_distrib': {
            'type': 'csv',
            'help': ('comma-separated list of distributions to build against'),
        },
        'lgp_sign': {
            'help': ('whether to sign packages'),
            'default': 'no'
        },
    }

    def do_check(self, test):
        dist = self.options.get('lgp_build_distrib') or ['all']
        sign = self.options.get('lgp_sign')
        cwd = os.getcwd()
        os.chdir(test.project_path())
        try:
                handler = LgpLogHandler(self.writer)
                cmd = LGP.get_command(self.command)
                cmd.logger.addHandler(handler)
                exit_status = cmd.main_run(['-v', '-s', sign, '-d', ','.join(dist), '-r', os.path.join(test.project_path(), '..')], LGP.rcfile)
                self.debian_changes = [FilePath(changes, type='debian.changes', distribution=os.path.basename(os.path.dirname(changes)))
                                       for changes in cmd.packages if changes.endswith('.changes')]
                if exit_status:
                    self.writer.fatal('lgp %s exited with status %s', self.command, exit_status)
                    self.set_status(apycot.ERROR)
                else:
                    test.done_steps.add(STEP_DEBIANPKG)
        finally:
            os.chdir(cwd)
        return handler.status

apycot.register('checker', LgpBuildChecker)
